beps/00010-event-auditor/README.md

```diff
@@ -0,0 +1,221 @@
+  status: 'failed';
+  errors: unknown[];
+};
+
+export type AuditEventStatus = AuditEventSuccessStatus | AuditEventFailureStatus;
+```
+
+#### EventAuditor Interface
+
+This interface defines the functionalities of an `EventAuditor` class. This class provides methods for:
+- Extracting the actor ID from an Express request (if available).
+- Creating detailed audit event information based on provided options.
+- Logging an audit event with a specific level (info, debug, warn, or error).
+
+```ts
+/**
+ * Common fields of an audit event. Note: timestamp and pluginId are automatically added at event creation.
+ *
+ * @public
+ */
+export type AuditEventDetails = {
+  actor: ActorDetails;
+  eventName: string;
+  stage: string;
+  request?: AuditRequest;
+  response?: AuditResponse;
+  meta: JsonValue;
+  isAuditEvent: true;
+} & AuditEventStatus;
+
+export type AuditEventDetailsOptions = {
+  eventName: string;
+  stage: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  actorId?: string;
+  request?: Request;
+} & (AuditEventSuccessStatus | AuditEventUnknownFailureStatus);
+
+export type AuditEventOptions = {
+  eventName: string;
+  message: string;
+  stage: string;
+  level?: 'info' | 'debug' | 'warn' | 'error';
+  actorId?: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  request?: Request;
+} & (AuditEventSuccessStatus | AuditEventUnknownFailureStatus);
+
+export type EventAuditorOptions = {
+  logger: LoggerService;
+  authService: AuthService;
+  httpAuthService: HttpAuthService;
+};
+
+export interface EventAuditor {
+  /**
+   * Processes an express request and obtains the actorId from it. Returns undefined if actorId is not obtainable.
+   *
+   * @public
+   */
+  getActorId(request?: Request): Promise<string | undefined>;
+
+  /**
+   * Generates the audit event details to place in the metadata argument of the logger
+   *
+   * Secrets in the metadata field and request body, params and query field should be redacted by the user before passing in the request object
+   * @public
+   */
+  createAuditEventDetails(
+    options: AuditEventDetailsOptions,
+  ): Promise<AuditEventDetails>;
+
+  /**
+   * Generates an Audit Event and logs it at the level passed by the user.
+   * Supports `info`, `debug`, `warn` or `error` level. Defaults to `info` if no level is passed.
+   *
+   * Secrets in the metadata field and request body, params and query field should be redacted by the user before passing in the request object
+   * @public
+   */
+  auditEvent(options: AuditEventOptions): Promise<void>;
+}
+```
+
+## Release Plan
+
+The release plan involves initially creating a shared audit event package. Following this, the audit event will be implemented in core packages and other plugins. The first targets should be high-priority areas, such as the scaffolder and catalog systems. Since adding the audit event will not disrupt existing functionality, the release plan is simplified.
+
+## Dependencies
```

Comment:
I'd prolly call out https://github.com/backstage/backstage/issues/24169 as a dependency too, if we want to be reliably grabbing the IP of the caller

Reply:
I agree, currently grabbing IP would only return the proxy IP if a proxy is being used to access the endpoints.

Reply:
How should I call out this dependency? I don't see an associated package.